#167316023 Secures the protected routes

sulenchy opened the pull request:
### What does this PR do?

- Secures the protected routes

#### Description of Task to be completed?

- secure  /api/v1/users/addusers route
- secure /api/v1/schools/create route
- secure /api/v1/results/addresults
- Write some test

#### How should this be manually tested?
- Clone the repo
- Run `npm i`
- Run `git pull`
- Checkout to the branch `ft-secure-protected-route-167316023`
- Run `npm test`

#### Any background context you want to provide?

- N/A

sulenchy committed:
ft(secure-protected-routes): secures the protected routes

- secure  /api/v1/users/addusers route
-secure /api/v1/schools/create route
- secure /api/v1/results/addresults

[Delivers #167316023]

jamsgra commented on server/routes/index.js:
```diff
@@ -17,8 +17,8 @@ router.get('/api/v1/users/logout', UserController.logout);
-router.post('/api/v1/results/addresults', AddResultsValidation.validateAddResults, ResultsController.addResults);
+router.post('/api/v1/users/addusers', checkUserIsLoggedIn, checkUserRole, AddUsersValidation.validateAddUsers, UserController.addUsers);
+router.post('/api/v1/schools/create', checkUserIsLoggedIn, checkUserRole, SchoolValidator.validateSchoolCreate, SchoolController.create);
+router.post('/api/v1/results/addresults', checkUserIsLoggedIn, checkUserRole, AddResultsValidation.validateAddResults, ResultsController.addResults);
```
Currently `checkUserRole` only allows `admin` & `super admin`, but the `addresults` route should also allow `teachers`

sulenchy committed:
ft(refactor-route): refactor the router

- add the adminPriviledgeRouter
- add the teacherPriviledgeRouter

[Delivers #167316023

jamsgra commented on server/middlewares/checkUser.js:
```diff
@@ -15,4 +15,36 @@ export const checkUserRole = (req, res,next) => {
+}
+
+
+export const checkTeacherAndAboveTole = (req, res,next) => {
```
Typo: checkTeacherAndAbove**Role**

Might be easier to name it: `checkTeacherPrivilege`, `checkAdminPrivilege` etc.

sulenchy replied:
Cool. Consider it done

jamsgra commented on server/routes/index.js:
```diff
@@ -1,24 +1,21 @@
+
+router.use(adminPriviledgeRouter);
+router.use(teacherPrivilegeRouter)
+
```
If we order it:

```
router.use(checkUserIsLoggedIn);
router.use(userRouter); // i.e dashboard, profile, settings etc.

router.use(checkTeacherPrivilege);
router.use(teacherPrivilegeRouter);

router.use(checkAdminPrivilege);
router.use(adminPrivilegeRouter);
```

then it means we don't have to insert the middleware in each individual route within the router. 

What do you think?

jamsgra replied:
We should also create middleware that checks whether a user has a `school_uid` to make sure that they have registered their school after signing up, and if not, should redirect them to the school registration page.

sulenchy replied:
> If we order it:
> 
> ```
> router.use(checkUserIsLoggedIn);
> router.use(userRouter); // i.e dashboard, profile, settings etc.
> 
> router.use(checkTeacherPrivilege);
> router.use(teacherPrivilegeRouter);
> 
> router.use(checkAdminPrivilege);
> router.use(adminPrivilegeRouter);
> ```
> 
> then it means we don't have to insert the middleware in each individual route within the router.
> 
> What do you think?

Cool. It makes the codebase looks cleaner.

sulenchy replied:
@jamsgra I have implemented the changes. Pls have a look at it

jamsgra replied:
Hey so I think we have to be careful with the order.

Any middleware we use will be applied to everything below it.

So for instance with: 

```
router.use(checkAdminPrivilege);
router.use(adminPrivilegeRouter);

router.use(checkTeacherPrivilege);
router.use(teacherPrivilegeRouter);
```
The `checkAdminPrivilege` will be applied to the `teacherPrivilegeRouter` as it is executed before that router, and so this will prevent teachers accessing those routes as they will be blocked by the `checkAdminPrivilege` middleware.

Instead if we order it this way:
```
router.use(checkTeacherPrivilege);
router.use(teacherPrivilegeRouter);

router.use(checkAdminPrivilege);
router.use(adminPrivilegeRouter);
```

The `checkAdminPrivilege` will only be applied to the `adminPrivilegeRouter`

sulenchy replied:
@jamsgra checkAdminPrivilege before checkTeacherPrivilege because checkIfUserHasSchool should not apply to `create school` API endpoint and it lives in the `adminPrivilege.js`

sulenchy committed:
ft(secure-protected-route): add checkIfUserhasSchool middleware

jamsgra commented on server/routes/teacherPriviledge.js:
```diff
@@ -0,0 +1,19 @@
+router.use(checkIfUserHasSchool);
+
+// get students from the database
+router.get('/api/v1/users/:query', UserController.getUsers);
```
A question about structuring the API routes:

Do you think this should be `/api/v1/users/getusers/:query` or we should leave it as `/api/v1/users/:query`? 

We will be adding a `getUser` route for the individual user, so I was wondering whether that would then be `/api/v1/users/getuser` or something along those lines

sulenchy replied:
@jamsgra Currently, the endpoint is /api/v1/users. We can still have /api/v1/user to get user profile(info).

sulenchy committed:
rebase off develop branch

jamsgra committed:
Ch: name change
